// wind-script/lib/run.js
"use strict";
const fs = require("fs");
const { colorConsole } = require("./logger");

Array.prototype.includesValue = function (value) {
    var includesValue = false
    for (let i = 0; i < this.length; i += 3) {
        includesValue = (this[i] === value);
        if (includesValue) return includesValue;
    }
    return includesValue;
}

var varNotError = true;
const varByType = {
    new: (type = "Str") => {
        varNotError = true;
        switch (type) {
            case "Num":
                return new Number();
            case "Str":
                return new String();
            case "Boo":
                return false;
            case "Date":
                return new Date();
            default:
                colorConsole.error(`Uncaught ReferenceError: ${type} is not a define type`);
                varNotError = false;
                return;
        }
    },
    var: (type = "Str", value) => {
        varNotError = true;
        switch (type) {
            case "Num":
                return Number(value);
            case "Str":
                return String(value);
            case "Boo":
                return value === "true";
            case "Date":
                return new Date(value);
            default:
                colorConsole.error(`Uncaught ReferenceError: ${type} is not a define type`);
                varNotError = false;
                return;
        }
    }
}
function dealWithVar(text) {
    if (text.includes("(") && text.slice(text.indexOf("}"), text.indexOf("(")).trim() === "}") {
        if (text.includes(")")) {
            let varType = text.slice(text.indexOf("{") + 1, text.indexOf("}"));
            let varValue = text.slice(text.indexOf("(") + 1, text.indexOf(")"));
            varValue = varByType.var(varType, varValue);
            if (varNotError) return [varType, varValue];
            else return [];

        }
        else {
            colorConsole.error("Uncaught SyntaxError: Unexpected token '('");
            return [];
        }
    }
    else {
        let varName = text.slice(text.indexOf("}") + 1, text.includes("=") ? text.indexOf("=") : text.length - 1).trim();
        let varType = text.slice(text.indexOf("{") + 1, text.indexOf("}"));
        let varValue = text.includes("=") ? text.slice(text.indexOf("=") + 1, text.length - 1).trim() : varByType.new(varType);
        varValue = varByType.var(varType, varValue);
        if (varNotError) return [varName, varType, varValue];
        else return [];
    }
}

class WSrun {
    constructor(debug = false) {
        this.debug = debug;
        return this;
    }
    setPath(path) {
        //读取文件
        var getFile = fs.readFileSync(path).toString();
        this.getCode = getFile
            .replace(/\r/g, "")
            .replace(/\n/g, "\\n");
        this.isCmd = false;
        return this;
    }
    setCode(code) {
        this.getCode = code;
        this.isCmd = false;
        return this;
    }
    setCmd(code) {
        this.getCode = code;
        this.isCmd = true;
        return this;
    }
    run() {
        //编译
        var getCode = this.getCode;
        var allCode = [""];
        var allVar = this.allVar || { data: [] };
        var noteText = "";
        for (let i = 0; i < getCode.length; i++) {
            if (noteText == "") {
                if (getCode[i] == ";") {
                    allCode[allCode.length - 1] += getCode[i];
                    allCode[allCode.length] = "";
                }
                else if (getCode[i] == "/") {
                    if (getCode[i + 1] == "/" || getCode[i + 1] == "*") {
                        noteText = `/${getCode[i + 1]}`;
                    }
                }
                else if (getCode[i] == "\\") {
                    if (getCode[i + 1] == "n") {
                        allCode[allCode.length] = "";
                        i++;
                    }
                }
                else {
                    allCode[allCode.length - 1] += getCode[i];
                }
            }
            else {
                if (noteText == "//") {
                    if (getCode[i] == "\\") {
                        if (getCode[i + 1] == "n") {
                            i++;
                            noteText = "";
                        }
                    }
                }
                if (noteText == "/*") {
                    if (getCode[i] == "*") {
                        if (getCode[i + 1] == "/") {
                            i++;
                            noteText = "";
                        }
                    }
                }
            }
        }
        allCode.forEach((value, i) => {
            allCode[i] = value.trim();
            if (allCode[i][allCode[i].length - 1] !== ";") {
                allCode[i] += ";";
            }
        });
        allCode.forEach((value, i) => {
            allCode[i] = value.trim();
            if (allCode[i] === ";") {
                allCode.splice(i, 1);
            }
        });

        //运行
        for (let i = 0; i < allCode.length; i++) {
            let code = allCode[i].trim();

            //var
            if (code.startsWith("{")) {
                if (code.includes("}")) {
                    var varDeal = dealWithVar(code);
                    if (varDeal.length != 0) {
                        if (!(allVar.data.includesValue(varDeal[0]))) {
                            allVar.data[allVar.data.length] = varDeal[0];
                        }
                        allVar.data[allVar.data.indexOf(varDeal[0]) + 1] = varDeal[1];
                        allVar.data[allVar.data.indexOf(varDeal[0]) + 2] = varDeal[2];
                        if (this.isCmd) colorConsole.log(varDeal[2]);
                    }
                }
                else {
                    colorConsole.error("Uncaught SyntaxError: Unexpected token '{'");
                }
            }
            this.allVar = allVar;

            //console
            if (code.includes("==>")) {
                var consoleUse = ["W warn", "E error", "L log", "I info", " log"];
                var consoleError = true;
                for (let i = 0; i < consoleUse.length; i++) {
                    if (code.startsWith(`${consoleUse[i].split(" ")[0]}==>`)) {
                        consoleError = false;
                        let logVar = code.slice(code.indexOf(`${consoleUse[i].split(" ")[0]}==>`) + (`${consoleUse[i].split(" ")[0]}==>`.length), code.length - 1).trim();
                        if (allVar.data.includesValue(logVar)) {
                            colorConsole[consoleUse[i].split(" ")[1]](allVar.data[allVar.data.indexOf(logVar) + 2]);
                        }
                        else {
                            colorConsole.error(`Uncaught ReferenceError: ${logVar} is not defined`);
                        }
                    }
                }
                if (consoleError) {
                    colorConsole.error(`Uncaught TypeError: ${code.slice(0, code.indexOf("==>"))} is not a console type`);
                }
            }
        }

        //debug
        if (this.debug) {
            console.log("getCode");
            console.log(getCode);
            console.log(`allCode: ${allCode.length}lines`);
            console.log(allCode);
            console.log(`allVar: ${allVar.length / 3} var`);
            console.log(allVar);
        }

        return this;
    }
}

exports.WSrun = WSrun;
